Enable replacing an object with a template

thabetx opened the pull request:
Solves #1722

Works exactly like the replace tile feature.

![replace_template](https://user-images.githubusercontent.com/7282243/31853827-c08b6cf4-b68f-11e7-883b-29e377bc5d5d.gif)

I had to move the selected template from the template creation tool into the tool manager so it can be used across all the tools, just like the tile.

bjorn commented on src/tiled/changemapobject.cpp:
```diff
@@ -245,19 +259,55 @@ void ResetInstances::undo()
+    for (int i = 0; i < mMapObjects.size(); ++i) {
+        MapObject *current = mMapObjects.at(i);
+        MapObject *old = mOldMapObjects.at(i);
+        copyProperties(old, current);;
```
Please remove second `;`

bjorn commented on src/tiled/changemapobject.cpp:
```diff
@@ -245,19 +259,55 @@ void ResetInstances::undo()
+
+
+ReplaceObjectsWithTemplate::ReplaceObjectsWithTemplate(MapDocument *mapDocument,
+                                                      const QList<MapObject *> &mapObjects,
```
Indentation missing a space.

bjorn commented on src/tiled/changemapobject.cpp:
```diff
@@ -245,19 +259,55 @@ void ResetInstances::undo()
     emit mMapDocument->selectedObjectsChanged();
 }
+
+ReplaceObjectsWithTemplate::~ReplaceObjectsWithTemplate()
```
Please define the destructor right after the constructor (matching the order in the header file).

bjorn commented on src/tiled/mapdocument.cpp:
```diff
@@ -139,7 +139,7 @@ void MapDocument::saveSelectedObject(const QString &name, int groupIndex)
         // Convert the saved object into an instance and clear the changed properties flags
-        object->setTemplateRef({objectTemplate->templateGroup(), objectTemplate->id()});
+        object->setTemplateRef(objectTemplate->templateRef());
         object->setChangedProperties(0);
```
Actually these lines should be covered by an undo command, right? Now that you've written `ReplaceObjectsWithTemplate`, you should probably use it here.

thabetx replied:
Hmm, currently this would be a partial undo, as removing from the template group is not implemented. But at least it would reset the object to being a normal object.

bjorn replied:
Yeah, I think partial undo would be fine here.

thabetx commented on src/tiled/abstractobjecttool.cpp:
```diff
@@ -249,6 +249,13 @@ void AbstractObjectTool::detachSelectedObjects()
+{
+    mapDocument()->undoStack()->push(new ReplaceObjectsWithTemplate(mapDocument(),
+                                                                    mapDocument()->selectedObjects(),
+                                                                    {objectTemplate()->templateGroup(), objectTemplate()->id()}));
```
For some reason I didn't commit that 😅.

thabetx commented on src/tiled/changemapobject.cpp:
```diff
@@ -204,6 +204,20 @@ void DetachObjects::undo()
     emit mMapDocument->mapObjectModel()->objectsChanged(mMapObjects);
 }
 
+void copyProperties(const MapObject *source, MapObject *target) {
```
Using the = operator looks like a good idea, it copies everything except id and position so I think i's okay but 'm not sure.

bjorn replied:
Hmm, I guess `operator=` could be expected to copy those as well, so probably better to leave this as a special function.

thabetx commented:
I like the option to add the name of the template in the context menu. It makes things clearer.

bjorn commented on src/tiled/changemapobject.cpp:
```diff
@@ -204,6 +204,20 @@ void DetachObjects::undo()
+    target->setShape(source->shape());
+    target->setCell(source->cell());
+    target->setRotation(source->rotation());
+    target->setVisible(source->VisibleProperty);
```
I just noticed this little issue. I won't change it now on master to avoid conflicts, so please fix this to `source->isVisible()` as part of this PR.

thabetx commented:
I cleaned the code and made some fixes and changes.

Replacing was enabled for tile objects only, so I fixed that as well, but now when replacing a non-tile object e.g. a polygon with a template, Tiled will crash. I think the reason is that the two objects have different sets of properties and when the properties dock is asked to update it can't set the properties of the new object.

So I will be looking into it and will do some more testing.

bjorn commented:
> Replacing was enabled for tile objects only, so I fixed that as well, but now when replacing a non-tile object e.g. a polygon with a template, Tiled will crash. I think the reason is that the two objects have different sets of properties and when the properties dock is asked to update it can't set the properties of the new object.

Ah, you're right. This new Replace by Template action is the first action that allows changing the actual type of an object, and some properties are only created for certain objects (like flipping for tile objects, and width/height only for rectangles, tiles and ellipses). So the PropertyBrowser will need some changes for this.

bjorn commented on src/tiled/abstractobjecttool.cpp:
```diff
@@ -358,13 +358,20 @@ void AbstractObjectTool::showContextMenu(MapObjectItem *clickedObjectItem,
+    auto replaceTemplateAction = menu.addAction(tr("Replace With Template"), this, SLOT(replaceObjectsWithTemplate()));
+
+    if (selectedTemplate)
+        replaceTemplateAction->setText(tr("Replace With \"") + selectedTemplate->name() + tr("\" Template"));
```
This is not nice for the translator. You should instead use a parameter, like so:
```c++
tr("Replace With \"%1\" Template").arg(selectedTemplate->name())
```

thabetx replied:
Alright.

bjorn commented:
Alright, I've merged `master` and in the process fixed that translation string and corrected `ReplaceObjectsWithTemplate::redo`.

The remaining issue is the crashing when changing the kind of object through replacing it with a template.

thabetx commented:
Hello bjron, thank you for the updates, I was stuck at updating the properties and the custom properties. I tested it and it's working great.

I will try working on #1823 .

Again thank you :).

bjorn commented:
> Hello bjorn, thank you for the updates, I was stuck at updating the properties and the custom properties. I tested it and it's working great.

No problem, I just needed this to be finished since I'm trying to move forward with a 1.1 release, which means finishing up the loose ends.

> I will try working on #1823.

Ah, I was about to look into it myself too, especially since I introduced this one recently, but feel free to have a look into it. It's probably not a hard one to fix.

thabetx commented:
Okay you can continue looking into it, I will try improving broken links.

Undoing after fixing doesn't reset and I will also look into fixing broken tileset link inside the template #1732 .
